src/java/org/apache/cassandra/db/ReadExecutionController.java

```diff
@@ -243,4 +261,27 @@ private void addSample()
         if (cfs != null)
             cfs.metric.topLocalReadQueryTime.addSample(cql, timeMicros);
     }
+
+    public void addActivationIds(ColumnFamilyStore.ViewFragment view)
+    {
+        Preconditions.checkState(metadata().replicationType().isTracked());
+        if (activationIds == null)
+            activationIds = new HashSet<>();
+        for (SSTableReader sstable : view.sstables)
+        {
+            ActivatedTransfers transfers = sstable.getCoordinatorLogOffsets().transfers();
+            if (transfers.isEmpty())
+                continue;
+            logger.trace("Adding overlapping IDs to read keyRange {}", command.dataRange.keyRange);
+            transfers.forEachIntersecting(command.dataRange.keyRange, id -> {
+                logger.debug("Adding overlapping activation ID {}", id);
+                activationIds.add(id);
+            });
+        }
+    }
+
+    public Iterator<ShortMutationId> getActivationIds()
+    {
+        return activationIds == null ? null : activationIds.iterator();
```

Comment:
Should we ever call `getActivationIds()` without having first set `activationIds`? If not, I would explode if we try to. I might also have `addActivationIds()` set `activationIds` to `Collections.empty()` or something as well.

Reply:
It's fine to have tracked reads with no intersecting transfers, so we shouldn't explode on unset